[Validation Tests]Add istio validation test of system-chart

guangbochen opened the pull request:
**Solution:**
This validation test will run the following tests against the rancher deployed istio services:
1. Run [dynamic admission webhook](https://istio.io/help/ops/setup/webhook/) prerequisites check of kube-apiserver.
2.  Verify the installed and deployed condition status of istio catalog app is `True`.
3. Verify istio(v1.1.5) CRD resources total count is equal to 53.
4. Run health-check of istio services.
5. Deploy bookinfo sample app and verify its workload status.
6. Deploy bookinfo gateway and verify its gateway's accessibility.


**Issue:**
https://github.com/rancher/rancher/issues/19087

**Related PRs:**
https://github.com/rancher/system-charts/pull/51

guangbochen commented:
**Verification Steps:**
1. Prepare either a singleton or HA-mode rancher server for validation test.
2. Add a new global custom catalog (e.g https://github.com/rancher/system-charts/pull/51) with [`rancher-istio` chart](https://github.com/guangbochen/system-charts/tree/istio-squash).
3. Config the following env parameters of the [validation tests](https://github.com/rancher/rancher/tree/master/tests/validation):
 > mostly you only need to update the `ISTIO_CATAGLOG` to your custom catalog name of istio ENV configs.
 
```
export ADMIN_TOKEN=<> && \
export CATTLE_TEST_URL=<> && \
export RANCHER_MACHINE_TIMEOUT=1200 && \
export ISTIO_CATAGLOG=system-library && \
export ISTIO_TEMPLATE=rancher-istio && \
export ISTIO_VERSION=1.1.5-rancher1 && \

pytest -v -s tests/validation/tests/v3_api/test_istio.py
```

loganhz commented:
LGTM

bmdepesa commented on tests/validation/tests/v3_api/test_istio.py:
```diff
@@ -0,0 +1,266 @@
+    return wl
+
+
+def create_and_verify_isito_app(p_client, ns, project, name):
```
istio* typo

bmdepesa commented on tests/validation/tests/v3_api/test_istio.py:
```diff
@@ -0,0 +1,266 @@
+    p_client.delete(workload)
+
+
+def add_isito_label_to_ns(c_client, ns):
```
istio* typo

bmdepesa commented on tests/validation/tests/v3_api/test_istio.py:
```diff
@@ -0,0 +1,266 @@
+namespace = {"p_client": None, "ns": None, "cluster": None, "project": None}
+istio_path = os.path.join(os.path.dirname(os.path.realpath(__file__)),
+                          "istio_files")
+ISTIO_CATAGLOG = os.environ.get('ISTIO_CATAGLOG', "system-library")
```
ISTIO_CATALOG* typo

bmdepesa replied:
In general, environment variables should be prefixed with `RANCHER_` so they can easily be parameterized from Jenkins -- all env vars prefixed with `RANCHER_` are collected and passed to the test container in Jenkins pipelines.

bmdepesa commented on tests/validation/tests/v3_api/test_istio.py:
```diff
@@ -0,0 +1,266 @@
+from .common import *  # NOQA
+
+namespace = {"p_client": None, "ns": None, "cluster": None, "project": None}
+istio_path = os.path.join(os.path.dirname(os.path.realpath(__file__)),
```
Resource files should be stored in `validation/tests/v3_api/resource` directory

bmdepesa commented on tests/validation/tests/v3_api/test_istio.py:
```diff
@@ -0,0 +1,266 @@
+def create_and_verify_isito_app(p_client, ns, project, name):
+    answers = {
+        "enableCRDs": "true",
+        "gateways.istio-ingressgateway.type": "NodePort"
```
Is it possible to specify the port? When running tests against our provisioned Azure cluster we will need to specifically open the port in the node template.

guangbochen commented:
updated to [279f1f1](https://github.com/rancher/rancher/compare/343576f423296920ce7e379c2c974de497e1a966..279f1f1d4b3c7bc9ad489b32bde5b2fdecaee5d3), @bmdepesa can you please help to review it, thanks.
* fixed typo and istio ENV naming
* moved istio files to `validation/tests/v3_api/resource` directory
* added nodePort env config with `RANCHER_ISTIO_INGRESSGATEWAY_NODEPORT`

bmdepesa commented:
LGTM

guangbochen commented:
Updated with a small fix of istio `auto-injection` ns label.

bmdepesa commented on tests/validation/tests/v3_api/test_istio.py:
```diff
@@ -0,0 +1,273 @@
+def verify_istio_crds_count():
+    get_crd_counts = execute_kubectl_cmd(
+        'get crds | grep "istio.io" | wc -l', False)
+    assert get_crd_counts == '53\n'
```
This value should be stored in a variable, and the comparison values should be stripped of whitespace beforehand. When running locally I get a failure due to whitespace mismatch in the string.

Why is this 53 in particular? Is this likely to change in the future?

guangbochen replied:
1. updated with set a variable and removed the trailing newline.
2. yes, it is committed to being 53 in `istio` v1.1.x versions, if we bump our istio chart version and the latest istio app update its CRD's [total count](https://istio.io/docs/setup/kubernetes/install/helm/), then we should update accordingly.

bmdepesa commented on tests/validation/tests/v3_api/test_istio.py:
```diff
@@ -0,0 +1,273 @@
+        name=name,
+        externalId=external_id,
+        targetNamespace=ns.name,
+        projectId=project.id,
```
Istio should be installed in the cluster's System project, as occurs when enabling manually from the UI. When istio is not deployed in the System project, graphs and metrics are not available within projects.

guangbochen replied:
updated, move the `istio apps` to be deployed in the `system` project and `istio-system` namespace.

bmdepesa commented:
@guangbochen can you update the CRD count validation and run the test against the latest Istio?

guangbochen commented:
let me check if we need to upgrade the istio version first before updating the validation test, will come back on this once the chart is done, thanks.

guangbochen commented:
Bump the test to be compatible with istio v1.3.0, but since we are planned to upgrade to istio v1.3.1, so will need to change the `ISTIO_VERSION` to whatever latest version is available in rancher/system-chart.

bmdepesa commented:
Closing in favor of https://github.com/rancher/rancher/pull/24468
